lib/CL/clBuildProgram.c

```diff
@@ -340,6 +341,41 @@ CL_API_SUFFIX__VERSION_1_0
   POCL_MSG_PRINT_INFO("building program with options %s\n",
                        program->compiler_options);
 
+  /* if we're rebuilding the program, release the kernels and reset log/status
+   */
+  if ((program->build_status != CL_BUILD_NONE) || program->num_kernels > 0)
+    {
+      cl_kernel k;
+      for (k = program->kernels; k != NULL; k = k->next)
+        {
+          k->program = NULL;
+          --program->pocl_refcount;
+        }
+      program->kernels = NULL;
+      if (program->num_kernels)
+        {
+          program->operating_on_default_kernels = 1;
```

Comment:
Actually i was mistaken, clCreateKernel does not lock at all when creating default kernels, so this is much easier to fix, just need to move `POCL_UNLOCK_OBJ(program);` line in clBuildProgram near the end.